src/Processor/Constant.php

```diff
@@ -79,4 +79,42 @@ public function getTokens()
     {
         return $this->tokens;
     }
+
+    /**
+     * Override processing of individual value.
+     *
+     * If the value is a string and evaluates to a class constant, returns
+     * the class constant value; otherwise, delegates to the parent.
+     *
+     * @param mixed $value
+     * @param array $replacements
+     * @return mixed
+     */
+    protected function doProcess($value, array $replacements)
+    {
+        if (! is_string($value)) {
+            return parent::doProcess($value, $replacements);
+        }
+
+        if (false === strpos($value, '::')) {
+            return parent::doProcess($value, $replacements);
+        }
+
+        // Handle class constants
+        if (defined($value)) {
+            return constant($value);
+        }
+
+        // Handle ::class notation
+        if (! preg_match('/::class$/', $value)) {
+            return parent::doProcess($value, $replacements);
+        }
+
+        $class = substr($value, 0, strlen($value) - 7);
+        if (class_exists($class)) {
+            return $class;
+        }
+
+        return parent::doProcess($value, $replacements);
```

Comment:
I decided to do some testing locally and see if it was possible to get to this point in the code in any meaningful way, and, well, it's not. If you were to try and define a constant named `Foo::class`, for instance, PHP tells you "Class contasnts cannot be defined or redefined", even if the class does not exist. Which means that once we get to this point, there's no way `$replacements` could even contain a value for it. As such, this should likely just be `return $value` — is that what you're suggesting, @Ocramius ?